### ISSUE TITLE
Values files in source control

### ISSUE THREAD

**Issue by AssafKatz3** [roboll/helmfile#29]
Hi,
It isn't clear how can put the values files inside source control (e.g. GIT). The best option will be allowing additional option in charts.yaml:
```
valueRepositories:
  - name: values
    url: http://roboll.io/values

```
Thanks

**Comment by mumoshu** [comment]
Hi @AssafKatz3,

Sry but I couldn't follow you!
What do you mean by putting values files under source control?

I believe committing `charts.yaml` along with `values.yaml` into the same git repo, and just running `helmfile sync` would work.
If that doesn't work for you, could you elaborate a bit more on your use-case?
Thanks!

**Comment by AssafKatz3** [comment]
Hi @mumoshu,
I have one repository which contains HELM charts of my software (micro services) and additional one which contains specific adoption for each enviornment.
Thanks

**Comment by mumoshu** [comment]
@AssafKatz3 Thanks for the clarification!
Then, I'd suggest putting your helmfile(=charts.yaml) to the environment repository like:

- environments/
  - prod/
    - charts.yaml
    - values.yaml
  - test/
    - charts.yaml
    - values.yaml

Your charts.yaml would then look like:

```yaml
repositories:
- name: yourcharts
  url: path/to/git-checked-out/charts

charts:
  name: myrelease
  chart: mychart
  values:
  -  values.yaml
```

**Comment by mumoshu** [comment]
@AssafKatz3 Hi! Did my answer make sense to you? In case I misunderstood your use-case, please let me know.

**Comment by AssafKatz3** [comment]
Hi,
Unfortunately, it didn’t helpful. We cannot put prod/ and test/ on same repository since they have totally different permission configuration, so I don’t see how it will helpful - when I run in production, I have permission to read test but not the other way around. So I cannot use this approach since each of file should come from another repository (most probably, from different GIT server as well).
Thanks

**Comment by cmeury** [comment]
@mumoshu As a side question: How do you handle clashing keys in your proposed pattern? AFAIK, unlike umbrella charts, helmfile does not currently support scoped values according to release name. Should it, potentially? If yes, I'll open a separate issue.

**Comment by cmeury** [comment]
@AssafKatz3 Would submodules be an option?

**Comment by mumoshu** [comment]
@AssafKatz3 I now believe I understand the usefulness of this feature more than before. But not yet sure how I could shape this, mainly due to possible edge-cases.

For example, how would you version-control your values from the repository?
Do we need to specify commit ID along with a repository URL not to consume new values unexpectedly?

```yaml
valueRepositories:
  - name: values
    url: http://roboll.io/values
    ref: yourcommitidhere
```

**Comment by mumoshu** [comment]
@cmeury It would be a good work-around!

If I become nit-picky, my only concern on the work-around is that the repository URL and the commit ID aren't obvious from just looking at helmfile. This might be a slight problem but does seem to spoil the original purpose of helmfile as a declarative spec of your stack.

**Comment by mumoshu** [comment]
@cmeury In which example do you see the key clashing issue?

In case we're talking about https://github.com/roboll/helmfile/issues/96#issuecomment-384641199, helmfile would just override previously occurred keys with latest values. But I remember @sstarcher wanted to rather fail-fast in such cases so that one won't unexpectedly override values.

**Comment by sstarcher** [comment]
I would recommend against doing something new like valueRepostory and instead treat the values entry as having a protocol
```
values:
- my/local/file.yaml
- git::https://bitbucket.org/foocompany/module_name.git?ref=hotfix
```

This would allow you to handle multiple source locations for values files.  An example of pre-existing art.

https://www.terraform.io/docs/modules/sources.html

**Comment by AssafKatz3** [comment]
@cmeury I don't understand how submodules will help, can you explain?

**Comment by AssafKatz3** [comment]
@mumoshu We need option to specificy the commit (can be tag, commit, whatever) in GIT. The idea of @sstarcher seems to be good, but we still need option to pass credentials somehow and need also support of SSH protocol.

**Comment by mumoshu** [comment]
@AssafKatz3 Regarding the submodules idea, probably you would just run `git clone --recursive` before `helmfile sync`, so that helmfile can consume any values.yaml from any submodule.

**Comment by mumoshu** [comment]
@sstarcher Sounds like a good idea!

If I add something, I would slightly prefer `git://github.com/mumoshu/foo#master` syntax seen in e.g. npm, and perhaps `s3://mybucket/prefix` in the future, so that it becomes regular URIs which is consumable with the standard `net/url.Parse`.

**Comment by AssafKatz3** [comment]
@mumoshu So I will point from prod/-->nft/-->dev/, I don't know if it is workable, but it seems worth a try.

**Comment by mumoshu** [comment]
@AssafKatz3 Probably I'm still missing something but I thought you'd just add `helmfile.yaml` to your project's main code-base, and required submodules matching prod, staging, dev env, etc.

This way, even though many ppl can checkout the main code-base, one without prod access would be unable to checkout the submodule hence your prod setup/credentials are kept private.

Does this make sense?

**Comment by AssafKatz3** [comment]
@mumoshu  I agree, the idea of submodules will solve the credentials problem and the location since each GIT repository will maintain this configuration. I will still need many helmfile.yaml, one for each step - prod, staging, dev, each of them load the previous steps and its own.
Better option, but more complex, will be support include option in helmfile itself, so I will have prod/helmfile.yaml that includes staging/helmfile.yaml (which in turn includes dev/helmfile.yaml).

**Comment by AssafKatz3** [comment]
@cmeury This isn't issue of restrict merge to master, the production doesn't have access to development master at all, only to its copy. I have team of production engineers can actually deploy to production, but they get some of infromation from previous steps (development, customization, ...). And yes, this is overcomplicating things, but I don't have better choice due to requirements.

**Comment by AssafKatz3** [comment]
From my point of view,  I got how the requirement should be implemented and only need #96 in order to continue so I close this issue
